daemon/containerd/image_delete.go

```diff
@@ -205,19 +238,68 @@ func sortParentsByAffinity(parents []imageWithRootfs) {
 	})
 }
 
-// isSingleReference returns true if there are no other images in the
-// daemon targeting the same content as `img` that are not dangling.
-func (i *ImageService) isSingleReference(ctx context.Context, img images.Image) (bool, error) {
-	refs, err := i.client.ImageService().List(ctx, "target.digest=="+img.Target.Digest.String())
-	if err != nil {
-		return false, err
+// getSameReferences returns the set of images which are the same as:
+// - the provided img if non-nil
+// - OR the first named image found in the provided image set
+// - OR the full set of provided images if no named references in the set
+//
+// References are considered the same if:
+// - Both contain the same name and tag
+// - Both contain the same name, one is untagged and no other differing tags in set
+// - One is dangling
+func (i *ImageService) getSameReferences(ctx context.Context, img *images.Image, imgs []images.Image) ([]images.Image, error) {
+	var (
+		named      reference.Named
+		tag        string
+		sameRef    []images.Image
+		digestRefs = []images.Image{}
+	)
+	if img != nil {
+		repoRef, err := reference.ParseNamed(img.Name)
+		if err != nil {
+			return nil, err
+		}
+		named = repoRef
+		if tagged, ok := named.(reference.Tagged); ok {
+			tag = tagged.Tag()
+		}
 	}
-	for _, ref := range refs {
-		if !isDanglingImage(ref) && ref.Name != img.Name {
-			return false, nil
+	for _, ref := range imgs {
+		if !isDanglingImage(ref) {
```

Comment:
This assumes that all `imgs` have the same `Target`, right? Should we mention that in the Godoc (or check that at the beginning and return error if the assumption is broken), or it's highly unlikely that this function might be used in some other case?